internal/controller/clusteripamclaim_controller.go

```diff
@@ -0,0 +1,262 @@
+	}
+	clusterIPAMSpec.ExternalIPClaims = claims
+	return nil
+}
+
+func (r *ClusterIPAMClaimReconciler) createOrUpdateIPClaims(ctx context.Context, ci *kcm.ClusterIPAMClaim, claims []corev1.ObjectReference, ipPool kcm.IPPoolSpec) ([]corev1.ObjectReference, error) {
+	desiredCount := ipPool.Count
+
+	if claims != nil {
+		currentCount := len(claims)
+		if currentCount > desiredCount {
+			// remove from the end of the list
+			removalList := claims[desiredCount:currentCount]
+			claims = claims[0:desiredCount]
+
+			for _, toRemove := range removalList {
+				err := r.Client.Delete(ctx, &ipamv1.IPAddressClaim{ObjectMeta: metav1.ObjectMeta{
+					Name:      toRemove.Name,
+					Namespace: toRemove.Namespace,
+				}})
+
+				if client.IgnoreNotFound(err) != nil {
+					return claims, fmt.Errorf("failed to delete IPAddressClaim: %s/%s", toRemove.Name, toRemove.Namespace)
+				}
+			}
+			// we've reached the desired count
+			desiredCount = 0
+		} else if currentCount <= desiredCount {
+			desiredCount -= currentCount
+		}
+	}
+
+	if ipPool.Name != "" {
+		for range desiredCount {
+			claim, err := r.CreateIPAddressClaim(ctx, ci.Namespace, ipPool.TypedLocalObjectReference, ci)
+			if err != nil {
+				return claims, fmt.Errorf("failed to create node ip address claims for: %s pool : %w", ipPool.Name, err)
+			}
+			claims = append(claims, claim)
+		}
+	}
+	return claims, nil
+}
+
+func (r *ClusterIPAMClaimReconciler) CreateIPAddressClaim(ctx context.Context, namespace string, pool corev1.TypedLocalObjectReference, ci *kcm.ClusterIPAMClaim) (corev1.ObjectReference, error) {
+	claim := ipamv1.IPAddressClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:    namespace,
+			GenerateName: pool.Name + "-",
+		},
+		Spec: ipamv1.IPAddressClaimSpec{
+			PoolRef: pool,
+		},
+	}
+
+	utils.AddOwnerReference(&claim, ci)
+	utils.AddLabel(&claim, kcm.OwnedByLabel, ci.Name)
+	if err := r.Create(ctx, &claim); err != nil {
+		return corev1.ObjectReference{Name: claim.Name, Namespace: namespace}, fmt.Errorf("failed to create IP address claim: %w", err)
+	}
+	return corev1.ObjectReference{Name: claim.Name, Namespace: namespace}, nil
+}
+
+func (r *ClusterIPAMClaimReconciler) Delete(ctx context.Context, ci *kcm.ClusterIPAMClaim) (ctrl.Result, error) {
+	l := ctrl.LoggerFrom(ctx)
+
+	l.Info("Removing Finalizer", "finalizer", kcm.ClusterIPAMClaimFinalizer)
+	if controllerutil.RemoveFinalizer(ci, kcm.ClusterIPAMClaimFinalizer) {
+		if err := r.Client.Update(ctx, ci); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to update ClusterIPAMClaim %s/%s: %w", ci.Namespace, ci.Name, err)
+		}
+	}
+	l.Info("ClusterIPAMClaim deleted")
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *ClusterIPAMClaimReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	r.defaultRequeueTime = 10 * time.Second
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&kcm.ClusterIPAMClaim{}).
```

Comment:
should the controller reconcile any events from its ownees e.g. `IPAddressClaim` or (if it will persist) `ClusterIPAM`?

Reply:
The initial question is still valid. And one extra: should we watch all of the events, e.g. `delete`?